### PR TITLE
With refactor

### PR DESCRIPTION
**Pull request by CyrusRoshan**

### CODE REVIEW

**Comment by tejasmanohar** [comment]
Exciting stuff

File: pkg/controller/service/with.go [hunk 1]
```diff
@@ -1,250 +1,541 @@
...
-				http.Error(w, err.Error(), http.StatusNotFound)
-				return
-			} else if err != nil {
-				log.WithError(err).Error("lookup role")
-				w.WriteHeader(http.StatusInternalServerError)
-				return
-			}
-			roleID = role.ID
-		}
+type FetchObject struct {
```
Comment (joshwget) [inline pkg/controller/service/with.go:23]: This isn't needed anymore?

File: pkg/controller/service/with.go [hunk 2]
```diff
@@ -1,250 +1,541 @@
...
+		log.WithError(err).Error("get project")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	f(project, device)
+}
+
+func (s *Service) withRole(w http.ResponseWriter, r *http.Request, project *models.Project, f func(role *models.Role)) {
+	if project == nil {
```
Comment (joshwget) [inline pkg/controller/service/with.go:230]: We could remove all these checks if we change `project *models.Project` -> `project models.Project`

File: pkg/controller/service/with.go [hunk 3]
```diff
@@ -1,250 +1,541 @@
...
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	f(application)
+}
 
-		handler(w, r, projectID, authenticatedUserID, authenticatedServiceAccountID, tokenID)
+func (s *Service) withRelease(w http.ResponseWriter, r *http.Request, project *models.Project, application *models.Application, f func(release *models.Release)) {
+	if application == nil || project == nil {
```
Comment (joshwget) [inline pkg/controller/service/with.go:449]: We could remove all these checks if we change `application *models.Application` -> `application models.Application`

File: pkg/controller/service/with.go (hunk 2 above)
Comment (CyrusRoshan) [reply pkg/controller/service/with.go:230]: True, but we frequently pass the structs down to multiple following middleware. It's kind of like our DB functions which only return nil if they also return an err, except in this case we return an http error and simply don't call the following handler. I'd actually be in favor of removing these checks entirely, for that reason.

**Pull request merged**